release-23.1: server/license: Integrate license throttle with SQL

spilchen opened the pull request:
Backport 1/1 commits from #130263.

/cc @cockroachdb/release

---

Prior to this, the throttling code had been implemented but wasn’t integrated with SQL. This change completes that integration, applying throttling only to external SQL queries—internal SQL queries are exempt. If throttling is active, an error is returned and propagated to the caller.

For licenses requiring telemetry, the latest telemetry ping time is used in the throttling calculation by implementing the
license.TelemetryStatusReporter interface.

Most throttling functionality remains disabled, enabled only in unit tests or via environment variables. Full throttling will be enabled in CRDB-41758.

Epic: CRDB-39988
Closes: CRDB-39991
Release note: None
Release justification: This work is part of the CockroachDB core deprecation.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/130872)
<!-- Reviewable:end -->

spilchen committed:
server/license: Integrate license throttle with SQL

Prior to this, the throttling code had been implemented but wasn’t
integrated with SQL. This change completes that integration, applying
throttling only to external SQL queries—internal SQL queries are exempt.
If throttling is active, an error is returned and propagated to the
caller.

For licenses requiring telemetry, the latest telemetry ping time is used
in the throttling calculation by implementing the
license.TelemetryStatusReporter interface.

Most throttling functionality remains disabled, enabled only in unit
tests or via environment variables. Full throttling will be enabled in
CRDB-41758.

Epic: CRDB-39988
Closes: CRDB-39991
Release note: None

rafiss reviewed (approved):
the backport lgtm. i left a few comments which we can discuss further, and if we want to do anything about them, fix them in a separate PR on master and backport separately.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130872#-:-O70gdf1AEniLaVe6fgJ:bp6wjl9)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/server/license/enforcer.go` line 295 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/130872#-O70eKmq7bcO3DrRAn7f:-O70eKmq7bcO3DrRAn7g:b-k1vw7o) ([raw file](https://github.com/cockroachdb/cockroach/blob/164440fb2c84e5cd26f8251a2d0e06b809be3160/pkg/server/license/enforcer.go#L295)):*
> ```Go
> 		if e.GetHasLicense() {
> 			err = errors.WithHintf(pgerror.Newf(pgcode.CCLValidLicenseRequired,
> 				"License expired. The maximum number of open transactions has been reached."),
> ```

the product brief (section 3.4) mentions that the expiration date should be part of this message. is that feasible?
___
*[`pkg/server/license/enforcer.go` line 299 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/130872#-O70ehtZBZarHqHrhmqz:-O70ehtZBZarHqHrhmr-:b-8zfgn8) ([raw file](https://github.com/cockroachdb/cockroach/blob/164440fb2c84e5cd26f8251a2d0e06b809be3160/pkg/server/license/enforcer.go#L299)):*
> ```Go
> 		} else {
> 			err = errors.WithHintf(pgerror.Newf(pgcode.CCLValidLicenseRequired,
> 				"No license installed. The maximum number of open transactions has been reached."),
> ```

nit: for this and the other error messages, would it make sense to say "maximum number of concurrently open transactions" ?

otherwise, it seems like we have a cap on total number of transactions one is allowed to open.
___
*[`pkg/server/license/enforcer.go` line 349 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/130872#-O70flC-3itlyrbuBWgt:-O70flC-3itlyrbuBWgu:bev9z5u) ([raw file](https://github.com/cockroachdb/cockroach/blob/164440fb2c84e5cd26f8251a2d0e06b809be3160/pkg/server/license/enforcer.go#L349)):*
> ```Go
> 
> 	gpEnd, _ := e.GetGracePeriodEndTS()
> 	log.Infof(ctx, "enforcer license updated: grace period ends at %q, telemetry required: %t",
> ```

perhaps this should also log when the license expires.

also, will this output be confusing for enterprise licenses? i believe it would show the unix epoch here.


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
Thanks for the feedback. I will address these separately in a PR against master.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130872#-:-O73suXp71snhAB3hOA-:b-4fzryu)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @spilchen)

___
*[`pkg/server/license/enforcer.go` line 295 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/130872#-O70eKmq7bcO3DrRAn7f:-O73s36n1deLa-y1WeD5:b-rtv9v6) ([raw file](https://github.com/cockroachdb/cockroach/blob/164440fb2c84e5cd26f8251a2d0e06b809be3160/pkg/server/license/enforcer.go#L295)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

the product brief (section 3.4) mentions that the expiration date should be part of this message. is that feasible?
</blockquote></details>

Yes, I think we include the expiry date in the message.
___
*[`pkg/server/license/enforcer.go` line 299 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/130872#-O70ehtZBZarHqHrhmqz:-O73s8i6-0fhil-ZROHN:b-dq2utz) ([raw file](https://github.com/cockroachdb/cockroach/blob/164440fb2c84e5cd26f8251a2d0e06b809be3160/pkg/server/license/enforcer.go#L299)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: for this and the other error messages, would it make sense to say "maximum number of concurrently open transactions" ?

otherwise, it seems like we have a cap on total number of transactions one is allowed to open.
</blockquote></details>

Yes, makes sense. I think that will make it less confusing.
___
*[`pkg/server/license/enforcer.go` line 349 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/130872#-O70flC-3itlyrbuBWgt:-O73sJPTCMTWafTW1pBv:b-yh7sm1) ([raw file](https://github.com/cockroachdb/cockroach/blob/164440fb2c84e5cd26f8251a2d0e06b809be3160/pkg/server/license/enforcer.go#L349)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

perhaps this should also log when the license expires.

also, will this output be confusing for enterprise licenses? i believe it would show the unix epoch here.
</blockquote></details>

Good suggestions. I think a separate message for when the license has no grace period makes sense.


<!-- Sent from Reviewable.io -->
